### ISSUE TITLE
Snapshot shipping is stuck on Satellite disconnected

### ISSUE THREAD

**Issue by kvaps** [LINBIT/linstor-server#169]
Ok, the same test from https://github.com/LINBIT/linstor-server/issues/168 but with the updated image:

```bash
# linstor s ship --from-node m1c10 --to-node m1c12 --resource one-image-32
ERROR:
    Connection to satellite 'm1c6' lost
SUCCESS:
    Suspended IO of 'one-image-32' on 'm1c10' for snapshot
SUCCESS:
    Suspended IO of 'one-image-32' on 'm1c12' for snapshot
SUCCESS:
    Suspended IO of 'one-image-32' on 'm1c10' for snapshot
SUCCESS:
    Suspended IO of 'one-image-32' on 'm1c12' for snapshot
SUCCESS:
    Aborted snapshot of 'one-image-32' on 'm1c10'
SUCCESS:
    Aborted snapshot of 'one-image-32' on 'm1c12'
WARNING:
Description:
    No active connection to satellite 'm1c6'
Details:
    The controller is trying to (re-) establish a connection to the satellite. The controller stored the changes and as soon the satellite is connected, it will receive this update.
SUCCESS:
    Resumed IO of 'one-image-32' on 'm1c10' after failed snapshot
SUCCESS:
    Resumed IO of 'one-image-32' on 'm1c12' after failed snapshot
```

```bash
# linstor n l 
╭───────────────────────────────────────────────────────╮
┊ Node  ┊ NodeType  ┊ Addresses               ┊ State   ┊
╞═══════════════════════════════════════════════════════╡
┊ m1c4  ┊ SATELLITE ┊ 10.28.36.164:3367 (SSL) ┊ Online  ┊
┊ m1c5  ┊ SATELLITE ┊ 10.28.36.165:3367 (SSL) ┊ Online  ┊
┊ m1c6  ┊ SATELLITE ┊ 10.28.36.166:3367 (SSL) ┊ Online  ┊
┊ m1c7  ┊ SATELLITE ┊ 10.28.36.167:3367 (SSL) ┊ Online  ┊
┊ m1c8  ┊ SATELLITE ┊ 10.28.36.168:3367 (SSL) ┊ Online  ┊
┊ m1c9  ┊ SATELLITE ┊ 10.28.36.169:3367 (SSL) ┊ Online  ┊
┊ m1c10 ┊ SATELLITE ┊ 10.28.36.170:3367 (SSL) ┊ Online  ┊
┊ m1c12 ┊ SATELLITE ┊ 10.28.36.172:3367 (SSL) ┊ Online  ┊
┊ pve1  ┊ SATELLITE ┊ 10.28.36.159:3367 (SSL) ┊ Online  ┊
┊ pve2  ┊ SATELLITE ┊ 10.28.36.160:3367 (SSL) ┊ Online  ┊
┊ pve3  ┊ SATELLITE ┊ 10.28.36.161:3367 (SSL) ┊ OFFLINE ┊
╰───────────────────────────────────────────────────────╯


# linstor s l -r one-image-32
╭────────────────────────────────────────────────────────────────────────────────────────────╮
┊ ResourceName ┊ SnapshotName ┊ NodeNames    ┊ Volumes  ┊ CreatedOn ┊ State                  ┊
╞════════════════════════════════════════════════════════════════════════════════════════════╡
┊ one-image-32 ┊ ship0        ┊ m1c10, m1c12 ┊ 0: 3 GiB ┊           ┊ Satellite disconnected ┊
┊ one-image-32 ┊ snapshot-0   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊           ┊ Successful             ┊
┊ one-image-32 ┊ snapshot-1   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊           ┊ Successful             ┊
┊ one-image-32 ┊ snapshot-2   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊           ┊ Successful             ┊
┊ one-image-32 ┊ snapshot-3   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊           ┊ Successful             ┊
╰────────────────────────────────────────────────────────────────────────────────────────────╯


# linstor v l  -r one-image-32
╭──────────────────────────────────────────────────────────────────────────────────────────────────────╮
┊ Node  ┊ Resource     ┊ StoragePool ┊ VolNr ┊ MinorNr ┊ DeviceName    ┊ Allocated ┊ InUse  ┊    State ┊
╞══════════════════════════════════════════════════════════════════════════════════════════════════════╡
┊ m1c10 ┊ one-image-32 ┊ thindata    ┊     0 ┊    1097 ┊ /dev/drbd1097 ┊  2.96 GiB ┊ Unused ┊ UpToDate ┊
┊ m1c12 ┊ one-image-32 ┊ thindata    ┊     0 ┊    1097 ┊ None          ┊  2.96 GiB ┊        ┊  Unknown ┊
┊ m1c6  ┊ one-image-32 ┊ thindata    ┊     0 ┊    1097 ┊ /dev/drbd1097 ┊  2.97 GiB ┊ Unused ┊ UpToDate ┊
╰──────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

```bash
root@m1c6:~# drbdadm status one-image-32
one-image-32 role:Secondary
  disk:UpToDate
  m1c10 role:Secondary
    peer-disk:UpToDate

root@m1c10:~# drbdadm status one-image-32
one-image-32 role:Secondary
  disk:UpToDate
  m1c6 role:Secondary
    peer-disk:UpToDate

root@m1c12:~# drbdadm status one-image-32
'one-image-32' not defined in your config (for this host).
```

**Comment by ghernadi** [comment]
Can you try simply deleting that snapshot?
A failed shipping will not be automatically retried, all you can do with it is trying to clean it up.

If you have configured periodical snapshot-shipping, the next attempt will not be effected of a previously failed shipment.

**Comment by kvaps** [comment]
After the deletion it seems now working:

```
root@linstor-controller-0:/# linstor s ship --from-node m1c10 --to-node m1c12 --resource one-image-32

SUCCESS:
    Suspended IO of 'one-image-32' on 'm1c10' for snapshot
SUCCESS:
    Suspended IO of 'one-image-32' on 'm1c12' for snapshot
SUCCESS:
    Suspended IO of 'one-image-32' on 'm1c6' for snapshot
SUCCESS:
    Suspended IO of 'one-image-32' on 'm1c10' for snapshot
SUCCESS:
    Suspended IO of 'one-image-32' on 'm1c12' for snapshot
SUCCESS:
    Took snapshot of 'one-image-32' on 'm1c10'
SUCCESS:
    Took snapshot of 'one-image-32' on 'm1c12'
SUCCESS:
    Resumed IO of 'one-image-32' on 'm1c10' after snapshot
SUCCESS:
    Resumed IO of 'one-image-32' on 'm1c6' after snapshot
SUCCESS:
    Resumed IO of 'one-image-32' on 'm1c12' after snapshot
SUCCESS:
    Resumed IO of 'one-image-32' on 'm1c12' after snapshot
SUCCESS:
    Resumed IO of 'one-image-32' on 'm1c10' after snapshot
```

```
╭──────────────────────────────────────────────────────────────────────────────────────────╮
┊ ResourceName ┊ SnapshotName ┊ NodeNames    ┊ Volumes  ┊ CreatedOn           ┊ State      ┊
╞══════════════════════════════════════════════════════════════════════════════════════════╡
┊ one-image-32 ┊ ship1        ┊ m1c10, m1c12 ┊ 0: 3 GiB ┊ 2020-08-11 07:31:21 ┊ Successful ┊
┊ one-image-32 ┊ ship2        ┊ m1c10, m1c12 ┊ 0: 3 GiB ┊ 2020-08-11 07:33:00 ┊ Successful ┊
┊ one-image-32 ┊ snapshot-0   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊                     ┊ Successful ┊
┊ one-image-32 ┊ snapshot-1   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊                     ┊ Successful ┊
┊ one-image-32 ┊ snapshot-2   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊                     ┊ Successful ┊
┊ one-image-32 ┊ snapshot-3   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊                     ┊ Successful ┊
╰──────────────────────────────────────────────────────────────────────────────────────────╯
```

But if at least one node will be inaccessible the operation will be failed

**Comment by kvaps** [comment]
I also bit confused that deactivated resource is shown as `Unknown`, I think it's better to rename it to something like `SnapTarget`  or etc.

```
# linstor r l -r one-image-32
╭───────────────────────────────────────────────────────────────────────────────╮
┊ ResourceName ┊ Node  ┊ Port ┊ Usage  ┊ Conns ┊    State ┊ CreatedOn           ┊
╞═══════════════════════════════════════════════════════════════════════════════╡
┊ one-image-32 ┊ m1c10 ┊ 7097 ┊ Unused ┊ Ok    ┊ UpToDate ┊                     ┊
┊ one-image-32 ┊ m1c12 ┊ 7097 ┊        ┊       ┊  Unknown ┊ 2020-08-10 14:15:27 ┊
┊ one-image-32 ┊ m1c6  ┊ 7097 ┊ Unused ┊ Ok    ┊ UpToDate ┊                     ┊
╰───────────────────────────────────────────────────────────────────────────────╯
```

**Comment by kvaps** [comment]
Also, if I remove the resource:
```bash
# linstor r d m1c12 one-image-32
INFO:
    Resource-definition property 'DrbdOptions/Resource/quorum' was removed as there are not enough resources for quorum
INFO:
    Resource-definition property 'DrbdOptions/Resource/on-no-quorum' was removed as there are not enough resources for quorum
SUCCESS:
Description:
    Node: m1c12, Resource: one-image-32 marked for deletion.
Details:
    Node: m1c12, Resource: one-image-32 UUID is: ff8c47c3-cb8b-4e95-a91e-339120a2b877
SUCCESS:
    Notified 'm1c10' that 'one-image-32' is being deleted on Node(s): [m1c12]
SUCCESS:
    Notified 'm1c6' that 'one-image-32' is being deleted on Node(s): [m1c12]
SUCCESS:
    Deleted 'one-image-32' on 'm1c12'
SUCCESS:
Description:
    Node: m1c12, Resource: one-image-32 deletion complete.
Details:
    Node: m1c12, Resource: one-image-32 UUID was: ff8c47c3-cb8b-4e95-a91e-339120a2b877
```


and try to create new resource but with `STORAGE` layer only

```bash
linstor r c m1c12 one-image-32 -l STORAGE
<Ctrl-C>
```

```
# linstor r l -r one-image-32
╭─────────────────────────────────────────────────────────────────────╮
┊ ResourceName ┊ Node  ┊ Port ┊ Usage  ┊ Conns ┊    State ┊ CreatedOn ┊
╞═════════════════════════════════════════════════════════════════════╡
┊ one-image-32 ┊ m1c10 ┊ 7097 ┊ Unused ┊ Ok    ┊ UpToDate ┊           ┊
┊ one-image-32 ┊ m1c12 ┊ 7097 ┊        ┊       ┊  Unknown ┊           ┊
┊ one-image-32 ┊ m1c6  ┊ 7097 ┊ Unused ┊ Ok    ┊ UpToDate ┊           ┊
╰─────────────────────────────────────────────────────────────────────╯
```

I found that the shipped snapshots are not deleted from `m1c12`:

```
# linstor s l -r one-image-32
╭──────────────────────────────────────────────────────────────────────────────────────────╮
┊ ResourceName ┊ SnapshotName ┊ NodeNames    ┊ Volumes  ┊ CreatedOn           ┊ State      ┊
╞══════════════════════════════════════════════════════════════════════════════════════════╡
┊ one-image-32 ┊ ship1        ┊ m1c10, m1c12 ┊ 0: 3 GiB ┊ 2020-08-11 07:31:21 ┊ Successful ┊
┊ one-image-32 ┊ ship2        ┊ m1c10, m1c12 ┊ 0: 3 GiB ┊ 2020-08-11 07:33:00 ┊ Successful ┊
┊ one-image-32 ┊ snapshot-0   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊                     ┊ Successful ┊
┊ one-image-32 ┊ snapshot-1   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊                     ┊ Successful ┊
┊ one-image-32 ┊ snapshot-2   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊                     ┊ Successful ┊
┊ one-image-32 ┊ snapshot-3   ┊ m1c4, m1c6   ┊ 0: 3 GiB ┊                     ┊ Successful ┊
╰──────────────────────────────────────────────────────────────────────────────────────────╯
```

**Comment by ghernadi** [comment]
There a a few things to reply to, so here we go :) 

> But if at least one node will be inaccessible the operation will be failed

Yes, I also noticed that one when trying to figure out what happened in your case.
I have to think about if I can fix this issue.

> I also bit confused that deactivated resource is shown as Unknown, I think it's better to rename it to something like SnapTarget or etc.

Yes, that is already fixed in the linstor-client. With the next client release the resource will not show as "Unknown" but as "Inactive"

> and try to create new resource but with STORAGE layer only

I somehow understand your intention, but I don't think I can allow this behaviour as this would require allowing mixing layer-lists within the same resource-definition.
Even if in your case the resource on m1c12 will effectively only use its STORAGE layer, it still has to know about the layers above it for example to correctly calculate the resulting size of the volume

> I found that the shipped snapshots are not deleted from m1c12:

Snapshots always survive the deletion of the resource they were taken of. This way you can still restore the snapshot into a new resource, even if the original resource was already removed

The other issues where the deletion is stuck etc is most likely due to the mixed layer-list. However, Linstor should not have allowed that, or at least report an error, but getting stuck
I will look into this.

**Comment by ghernadi** [comment]
Update: Apparently I was wrong regarding mixing layer-lists.

It is indeed allowed, however your scenario is still not recommended.

The problem is that the snapshot (on both, sender and receiver) will also remember the layer-list of their original resource. That means, even if one can argue that shipping a DRBD-resource into a STORAGE-only-resource makes sense, the snapshot on the receiver site also stored the STORAGE only layer-list. Which means, you have saved your data, but you cannot (easily) restore it into a new resource as that new resource would also become a STORAGE only, not a DRBD resource as the sender of the snapshot shipment.

However, I still have to look into this, as the command should definitely not get stuck.

**Comment by ghernadi** [comment]
I tried reproducing this for a while, and although this is not recommended thing to do, it worked just fine. I was able to:
* create 2 DRBD-resources
  * create a third STORAGE-only
  * delete the STORAGE-only resource
* create 3 DRBD-resources (new resource-definition)
  * delete one of them and recreate it as a STORAGE-only

Both scenarios with and without leaving the deleted resource as a tie-breaker and recreating the STORAGE-only resource "on top of tiebreaker" (although also this is definitely not recommended)

**Comment by kvaps** [comment]
Hi, I'm not sure, but try this scenario:

1. disable auto tiebreaker
2. create DRBD resource on `node1`, `node2` and wait until synchronization is finished
3. create DRBD resource on `node3`
4. deactivate resource on `node3`
5. ship snapshot from `node1` to `node3`
6. remove resource on `node3`
7. create STORAGE resource on `node3`

**Comment by ghernadi** [comment]
As the client command gets stuck, you do not get any ErrorReports presented even if they are created... can you please check if there are some new ErrorReports after a while of stuck client?

If you get something like
```
Error message:                      The local resource with the UUID '8b10145b-a7dc-46b9-87c9-e1c010b32794' was not found in the stored resource definition 'rsc'.
```
then I found the error :)

**Comment by kvaps** [comment]
For now I can't reproduce, but I think I found another unwanted behavior.

when I try to place new resource with `-l STORAGE` on some node without storage pool specified I'm getting an error:

```bash
# linstor r c m1c12 testss -l STORAGE
ERROR:
Description:
    The storage pool 'DfltStorPool' for resource 'testss' for volume number '0' is not deployed on node 'm1c12'.
Cause:
    Dependency not found
Details:
    The resource which should be deployed had at least one volume definition (volume number '0') which LinStor tried to automatically create. The storage pool name for this new volume was looked for in order in the properties of the resource, volume definition, resource definition and node, and finally in a system wide default storage pool name defined by the LinStor controller.
    Node: m1c12, Resource: testss
Show reports:
    linstor error-reports show 5F324F93-00000-000001
```

[5F324F93-00000-000001.log](https://github.com/LINBIT/linstor-server/files/5062193/5F324F93-00000-000001.log)

but in the end resource is created:

```
# linstor r l -r testss
╭───────────────────────────────────────────────────────────────────────────────────────────╮
┊ ResourceName ┊ Node  ┊ Port ┊ Usage  ┊ Conns             ┊    State ┊ CreatedOn           ┊
╞═══════════════════════════════════════════════════════════════════════════════════════════╡
┊ testss       ┊ m1c12 ┊ 7533 ┊        ┊                   ┊  Unknown ┊ 2020-08-12 10:47:59 ┊
┊ testss       ┊ m1c4  ┊ 7533 ┊ Unused ┊ Connecting(m1c12) ┊ UpToDate ┊ 2020-08-12 10:47:06 ┊
┊ testss       ┊ m1c5  ┊ 7533 ┊ Unused ┊ Connecting(m1c12) ┊ UpToDate ┊ 2020-08-12 10:47:07 ┊
╰───────────────────────────────────────────────────────────────────────────────────────────╯
```

**Comment by kvaps** [comment]
Probably fixed in 1.8.0
